docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,226 @@
+---
+title: Insulate users from kubeadm API version changes
+authors:
+- "@fabriziopandini"
+reviewers:
+- "@vincepri"
+creation-date: 2021-02-10
+last-updated: 2021-02-10
+status: implementable
+see-also:
+- "/docs/proposals/20190610-machine-states-preboot-bootstrapping.md"
+replaces:
+superseded-by:
+---
+
+# Insulate users from kubeadm API version changes
+
+## Table of Contents
+
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Make CABPK and KCP to use more recent versions of the kubeadm API and insulate users from
+kubeadm API version changes.
+
+## Motivation
+
+Cluster bootstrap provider for kubeadm (CABPK) and the control plane provider for kubeadm
+(KCP) API still relying on kubeadm v1beta1 API, which has been deprecated, and it is going
+to be removed ASAP.
+
+While moving to a more recent version of the kubeadm API is a required move, Cluster API
+should take this opportunity stop to exposing the kubeadm API types in the
+KubeadmConfig/KubeadmControlPlane specs, and instead use our own types.
```

Comment:
i think embedding the kubeadm v1beta1 type was a mistake.

a non-v1 type should not be embedded in another type as as a field. but there is also a risk for v1 types in a dynamic area of k8s. if v2 comes out (inevitable in some places) you'd have to bump your API if you want to use the v2 features.

wrapping kubeadm types will inevitably introduce burden on CAPI and just more types to maintain. how much are you going to expose out of the kubeadm API?

i think the right pattern is to:
- pass the kubeadm types as separate structures (separated with `---\n`)
- let users configure everything
- document what cannot be configured and CAP* will override it
- error out on outdated APIs

Reply:
There is a little bit of confusion here, I rephrased it to make it more clear, PTAL.